### PR TITLE
Merge wellspring back upstream

### LINKED ISSUE #212: Fails to build with gcc 10
gcc 10 has landed in Fedora Rawhide.  gfortran 10 now enforces type matching in at least some function calls.  netcdf-fortran 4.5.2 fails with:
```
netcdf4_func.f90:730:75:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/REAL(8)).
netcdf4_func.f90:720:75:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(8)/REAL(8)).
netcdf4_func.f90:710:74:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
netcdf4_func.f90:700:73:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(2)/REAL(8)).
netcdf4_func.f90:690:73:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(1)/REAL(8)).
netcdf4_func.f90:670:75:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/REAL(8)).
netcdf4_func.f90:660:75:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(8)/REAL(8)).
netcdf4_func.f90:650:74:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(8)).
netcdf4_func.f90:640:73:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(2)/REAL(8)).
netcdf4_func.f90:630:73:
......
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(1)/REAL(8)).
```

### LINKED ISSUE #218: Build fails with NAG compiler 6.2
* 4.4.4 and 4.5.2
* CentOS7, NAG 6.2 Build 6223, GCC 7.3.0, netCDF 4.6.1 (no MPI), HDF 1.10.2 (no MPI)
* I think https://github.com/Unidata/netcdf-fortran/issues/212 has the same issue.
* Reproducible via `export FC=nagfor && ./configure && make`

Error:

nagfor -I. -g -c -o module_netcdf4_f03.o  module_netcdf4_f03.f90
NAG Fortran Compiler Release 6.2(Chiyoda) Build 6223
[NAG Fortran Compiler normal termination]
nagfor -I. -g -c -o netcdf4.o  netcdf4.f90
NAG Fortran Compiler Release 6.2(Chiyoda) Build 6223
Warning: netcdf_text_variables.f90, line 19: Unused dummy variable MAP
Warning: netcdf_text_variables.f90, line 39: Unused dummy variable MAP
Warning: netcdf4_func.f90, line 761: Unused dummy variable LENGTH
Warning: netcdf4_func.f90, line 781: Unused dummy variable MAP
Warning: netcdf4_func.f90, line 801: Unused dummy variable MAP
Warning: netcdf4.f90, line 67: Unused external reference NF_FREE_STRING
Warning: netcdf4.f90, line 67: Unused external reference NF_FREE_VLENS
Warning: netcdf4.f90, line 67: Unused external reference NF_GET_VAR1_TEXT
Warning: netcdf4.f90, line 67: Unused external reference NF_GET_VARA_TEXT
Warning: netcdf4.f90, line 67: Unused external reference NF_GET_VAR_INT64
Warning: netcdf4.f90, line 67: Unused external reference NF_INQ_VAR_SZIP
Warning: netcdf4.f90, line 67: Unused external reference NF_PUT_VAR1_TEXT
Warning: netcdf4.f90, line 67: Unused external reference NF_PUT_VARA_TEXT
Warning: netcdf4.f90, line 67: Unused external reference NF_PUT_VAR_INT64
Warning: netcdf4.f90, line 67: Unused external reference NF_PUT_VLEN_ELEMENT
Error: netcdf4_func.f90, line 640: Inconsistent data type INTEGER(int16) (previously INTEGER(int8)) for argument 4 in reference to NF_DEF_VAR_FILL
Error: netcdf4_func.f90, line 650: Inconsistent data type INTEGER (previously INTEGER(int8)) for argument 4 in reference to NF_DEF_VAR_FILL
Error: netcdf4_func.f90, line 660: Inconsistent data type INTEGER(int64) (previously INTEGER(int8)) for argument 4 in reference to NF_DEF_VAR_FILL
Error: netcdf4_func.f90, line 670: Inconsistent data type REAL (previously INTEGER(int8)) for argument 4 in reference to NF_DEF_VAR_FILL
Error: netcdf4_func.f90, line 680: Inconsistent data type DOUBLE PRECISION (previously INTEGER(int8)) for argument 4 in reference to NF_DEF_VAR_FILL
Error: netcdf4_func.f90, line 700: Inconsistent data type INTEGER(int16) (previously INTEGER(int8)) for argument 4 in reference to NF_INQ_VAR_FILL
Error: netcdf4_func.f90, line 710: Inconsistent data type INTEGER (previously INTEGER(int8)) for argument 4 in reference to NF_INQ_VAR_FILL
Error: netcdf4_func.f90, line 720: Inconsistent data type INTEGER(int64) (previously INTEGER(int8)) for argument 4 in reference to NF_INQ_VAR_FILL
Error: netcdf4_func.f90, line 730: Inconsistent data type REAL (previously INTEGER(int8)) for argument 4 in reference to NF_INQ_VAR_FILL
Error: netcdf4_func.f90, line 740: Inconsistent data type DOUBLE PRECISION (previously INTEGER(int8)) for argument 4 in reference to NF_INQ_VAR_FILL
[NAG Fortran Compiler error termination, 10 errors, 15 warnings]
make[1]: *** [netcdf4.o] Error 2
make[1]: Leaving directory `/gpfs0/home/admtost/builds/netcdf-fortran-4.5.2/fortran'
make: *** [all-recursive] Error 1

### LINKED ISSUE #249: We should check that automake is of sufficient version (1.13 at least)
We should be doing:

`AM_INIT_AUTOMAKE([1.13 foreign])`

So that we don't get users having this problem:
`acinclude.m4:403: warning: macro `AM_EXTRA_RECURSIVE_TARGETS' not found in library`

### LINKED ISSUE #252: CMake Build INCLUDEDIR not set properly
When using the CMake build, the `Fortran_INSTALL_MODDIR` variable is set to a Boolean value and later on used to determine a path (in `fortran/CMakeLists.txt`):
```
INSTALL(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/
        DESTINATION "${Fortran_INSTALL_MODDIR}"
        FILES_MATCHING PATTERN "*.mod" PATTERN "netcdf.inc"
        )
``` 

So I am wondering if we should replace in `CMakeLists.txt.`:
```
OPTION(Fortran_INSTALL_MODDIR "Fortran module install directory" "${CMAKE_INSTALL_INCLUDEDIR}")
```
by
```
SET(Fortran_INSTALL_MODDIR "${CMAKE_INSTALL_INCLUDEDIR}" CACHE STRING "Fortran module install directory")
```

Just for my understanding, why is there a variable `Fortran_INSTALL_MODDIR` that is used instead of `CMAKE_INSTALL_INCLUDEDIR`?

### PR DESCRIPTION
**Pull request by WardF**
* Fixes #249 
* Fixes #252 
* Fixes #212
* Fixes #218 

Branch incorporating changes for upcoming version that will also be merged back in to `master` when the new version is released.

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -5,8 +5,13 @@ Release Notes {#nf_release_notes}
...
 This file contains a high-level description of this package's evolution.
 Entries are in reverse chronological order (most recent first).
+## 4.5.4 - TBD
 
-## 4.5.3 - TBD
+### Requirements
+
+* netCDF-C: 4.7.4+
+
+## 4.5.3 - April 24, 2020
```
Comment (MuellerSeb) [inline RELEASE_NOTES.md:14]: Where does this date come from?

Comment (WardF) [reply RELEASE_NOTES.md:14]: It was the original target date, but we have been overtaken by circumstances and will be updated when the 4.5.3 release is actually minted.

File: CMakeLists.txt [hunk 2]
```diff
@@ -545,6 +547,18 @@ ENDIF()
 # checks, like in configure.ac.
 ###
 
+##
+# Check to see if compiler supports
+# -fallow-argument-mismatch-flag introduced in gcc 10.
+#
+# See https://github.com/Unidata/netcdf-fortran/issues/212
+##
+check_Fortran_compiler_flag("-fallow-argument-mismatch" COMPILER_HAS_ALLOW_ARGUMENT_MISMATCH)
```
Comment (MuellerSeb) [inline CMakeLists.txt:556]: For the NAG compiler, it is `-mismatch_all`, as written here: https://github.com/Unidata/netcdf-fortran/issues/218

Comment (WardF) [reply CMakeLists.txt:556]: Excellent, I'll add a similar check!

**Comment by WardF** [comment]
Added a fix for NAG compilers, if somebody with access to NAG could check that and let me know for certain, I'd appreciate it :).  Otherwise, I think we're ready to proceed with the release and then can merge all of this back in to `master`.

**Pull request merged**